File: control-server/src/services/accountInsights.ts
import type { User } from "@prisma/client"
import { config } from "../config"
import { effectiveDeviceLimit } from "../lib/deviceLimit"
import { bytesToNumber, prisma } from "../prisma"
import { egressBudgetView } from "./egressBudget"
import { lookupOrigin } from "./geo"
import { countryPoint, usageBucket, usageWindow, type UsagePeriod } from "./insightMath"
import { loadPublicNodes } from "./nodes"
import { serviceSettings, serviceStatus } from "./serviceControl"

type Origin = { lat: number; lon: number; countryCode: string | null; country: string | null; source: "ip-country" | "device-estimate"; approximate: true }
// Cache by session, not account or caller IP. A phone and PC must never inherit
// each other's position. No IPs or credentials appear in the response or cache key.
const origins = new Map<string, { expires: number; value: Promise<Origin | null> }>()
async function sessionOrigin(id: string, ip: string | null): Promise<Origin | null> {
	if (!ip || !config.GEOIP_ENABLED) return null
	const cached = origins.get(id)
	if (cached && cached.expires > Date.now()) return cached.value
	if (origins.size >= 10000) origins.delete(origins.keys().next().value as string)
	const value = lookupOrigin(ip).then((geo): Origin | null => {
		const point = countryPoint(geo?.countryCode)
		return point && geo ? { lat: point[0], lon: point[1], countryCode: geo.countryCode, country: geo.country, source: "ip-country", approximate: true } : null
	})
	const entry = { expires: Date.now() + 3600000, value }
	origins.set(id, entry)
	const result = await value
	if (!result) entry.expires = Date.now() + 60000
	return result
}
export function deviceEstimate(code: string | null | undefined): Origin | null {
	const point = countryPoint(code)
	return point ? { lat: point[0], lon: point[1], countryCode: String(code).toUpperCase(), country: null, source: "device-estimate", approximate: true } : null
}

// Иконка сайта — только для настоящих доменов.
//
// Раньше адрес иконки склеивался безусловно, поэтому для сеансов без SNI,
// где «домен» — это IP, получалось
// https://icons.duckduckgo.com/ip3/77.111.249.147.ico. Такого файла не
// существует и не может существовать — сервис отвечает 404, а клиенты
// честно сообщали об этом в телеметрию (`NetworkImageLoadException:
// HTTP 404`). У IP иконки нет — отдаём null, и интерфейс ставит свою
// заглушку вместо заведомо битой картинки.
const IPV4_LITERAL = /^\d{1,3}(?:\.\d{1,3}){3}$/
export function domainFaviconUrl(domain: string): string | null {
	const host = String(domain ?? "").trim().toLowerCase().replace(/\.$/, "")
	if (!host || host.length > 253) return null
	// IPv6 и порты — тоже не домены.
	if (host.includes(":") || IPV4_LITERAL.test(host)) return null
	// Должна быть хотя бы одна точка и буквенный TLD.
	if (!/^[a-z0-9.-]+$/.test(host) || !/\.[a-z]{2,}$/.test(host)) return null
	return ["https:", "", "icons.duckduckgo.com", "ip3", encodeURIComponent(host) + ".ico"].join("/")
}

export async function recordMapCountry(userId: string, deviceId: string, countryCode: string) {
	return prisma.device.updateMany({
		where: { userId, id: deviceId, status: "ACTIVE", OR: [{ mapCountryCode: null }, { mapCountryCode: { not: countryCode } }] },
		data: { mapCountryCode: countryCode },
	})
}

export async function accountActiveMap(user: User, currentDeviceId?: string) {
	const [sessions, service] = await Promise.all([
		prisma.session.findMany({
			where: { userId: user.id, status: { in: ["PENDING", "ACTIVE"] }, device: { status: "ACTIVE" } },
			include: { device: true, node: true }, orderBy: { connectedAt: "desc" },
		}), serviceStatus(),
	])
	const unique = [...new Map(sessions.slice().reverse().map((s) => [s.deviceId, s])).values()].sort((a, b) => b.connectedAt.getTime() - a.connectedAt.getTime())
	const selected = unique
	const nodes = await loadPublicNodes(selected.map((s) => s.node))
	const now = new Date()
	const devices = await Promise.all(selected.map(async (s, i) => ({
		id: s.deviceId, deviceName: s.device.deviceName, platform: s.device.platform,
		lastSeen: s.device.lastSeen?.toISOString() ?? null, isCurrent: currentDeviceId === s.deviceId,
		connected: s.status === 'ACTIVE', sessionId: s.id, status: s.status, connectedAt: s.connectedAt.toISOString(),
		durationSec: Math.max(0, Math.floor((now.getTime() - s.connectedAt.getTime()) / 1000)),
		origin: (await sessionOrigin(s.id, s.clientIp)) ?? deviceEstimate(s.device.mapCountryCode), node: nodes[i],
	})))
	return { serverTime: now.toISOString(), pollAfterMs: 5000, activeTunnels: selected.filter(s => s.status === 'ACTIVE').length, pendingTunnels: selected.filter(s => s.status === 'PENDING').length, maxDevices: effectiveDeviceLimit(user), truncated: unique.length > selected.length, service, devices }
}

type Counters = { downloadBytes: number; uploadBytes: number }
const emptyCounters = (): Counters => ({ downloadBytes: 0, uploadBytes: 0 })
const totalBytes = (v: Counters) => v.downloadBytes + v.uploadBytes
export async function accountAnalytics(userId: string, period: UsagePeriod, now = new Date(), includeBudget = false) {
	const window = usageWindow(period, now)
	const domainCutoff = new Date(now.getTime() - config.DOMAIN_STATS_RETENTION_DAYS * 86400000)
	const [rows, domains, budget, settings] = await Promise.all([
		prisma.trafficUsageBucket.findMany({ where: { userId, bucketStart: { gte: window.start, lte: now } }, orderBy: { bucketStart: "asc" } }),
		config.DOMAIN_STATS_ENABLED ? prisma.trafficDomainStat.groupBy({
			by: ["domain", "category"], where: { userId, lastSeenAt: { gte: domainCutoff } },
			_sum: { bytesRx: true, bytesTx: true, connections: true }, _max: { lastSeenAt: true },
		}) : [],
		includeBudget ? egressBudgetView(now) : null, serviceSettings(),
	])
	const totals = emptyCounters()
	const series = new Map<string, { start: string } & Counters>()
	// Zero is only used inside the known observation coverage, never for a day
	// before installation. A partial first hour/day is explicitly marked below.
	const since = settings.analyticsSince ? new Date(settings.analyticsSince) : now
	const first = new Date(Math.max(window.start.getTime(), since.getTime()))
	const step = window.bucketSize === "hour" ? 3600000 : 86400000
	for (let t = new Date(usageBucket(first, window.bucketSize)).getTime(); t <= now.getTime(); t += step) {
		const start = new Date(t).toISOString()
		series.set(start, { start, ...emptyCounters() })
	}
	const devices = new Map<string, { deviceId: string; deviceName: string; platform: string | null } & Counters>()
	for (const row of rows) {
		const downloadBytes = bytesToNumber(row.downloadBytes)
		const uploadBytes = bytesToNumber(row.uploadBytes)
		totals.downloadBytes += downloadBytes; totals.uploadBytes += uploadBytes
		const key = usageBucket(row.bucketStart, window.bucketSize)
		const bucket = series.get(key) ?? { start: key, ...emptyCounters() }
		bucket.downloadBytes += downloadBytes; bucket.uploadBytes += uploadBytes
		series.set(key, bucket)
		const device = devices.get(row.deviceId) ?? { deviceId: row.deviceId, deviceName: row.deviceName, platform: row.platform, ...emptyCounters() }
		device.downloadBytes += downloadBytes; device.uploadBytes += uploadBytes
		device.deviceName = row.deviceName; device.platform = row.platform
		devices.set(row.deviceId, device)
	}
	const categories = new Map<string, { category: string } & Counters>()
	const domainItems = domains.map((row) => {
		const value = { domain: row.domain, category: row.category ?? "other", downloadBytes: bytesToNumber(row._sum.bytesTx), uploadBytes: bytesToNumber(row._sum.bytesRx), connections: row._sum.connections ?? 0, lastSeenAt: row._max.lastSeenAt?.toISOString() ?? null, faviconUrl: domainFaviconUrl(row.domain) }
		const category = categories.get(value.category) ?? { category: value.category, ...emptyCounters() }
		category.downloadBytes += value.downloadBytes; category.uploadBytes += value.uploadBytes
		categories.set(value.category, category)
		return value
	}).sort((a, b) => totalBytes(b) - totalBytes(a)).slice(0, 20)
	return {
		period, start: window.start.toISOString(), end: now.toISOString(), bucketSize: window.bucketSize,
		coverage: { since: settings.analyticsSince, partial: !settings.analyticsSince || since > window.start, source: "session-counter-deltas", timezone: "UTC" },
		totals, series: [...series.values()].sort((a, b) => a.start.localeCompare(b.start)),
		devices: [...devices.values()].sort((a, b) => totalBytes(b) - totalBytes(a)),
		domains: { enabled: config.DOMAIN_STATS_ENABLED, windowDays: config.DOMAIN_STATS_RETENTION_DAYS, scope: "retained-session-totals", items: domainItems },
		categories: [...categories.values()].sort((a, b) => totalBytes(b) - totalBytes(a)),
		// The OCI allowance belongs to the SERVICE, not every subscription. Cost,
		// error strings, OCI resource identifiers and other users are never exposed.
		budget: budget ? { adminOnly: true, scope: "service", source: "oci", available: budget.configured && Boolean(budget.lastPolledAt) && !budget.lastError,
			usedBytes: budget.usedBytes, budgetBytes: budget.budgetBytes, usedPercent: budget.usedPercent,
			cycleStart: budget.cycleStart, cycleEnd: budget.cycleEnd, lastPolledAt: budget.lastPolledAt } : null,
	}
}

File: flutter-client/lib/models/account_insights.dart
library;

import 'models.dart';

Map<String, dynamic> _map(Object? value) => value is Map
    ? value.map((Object? k, Object? v) => MapEntry<String, dynamic>('$k', v))
    : const <String, dynamic>{};
int _int(Object? value) => value is num && value.isFinite ? value.toInt() : int.tryParse('$value') ?? 0;
double? _double(Object? value) { final parsed = value is num ? value.toDouble() : double.tryParse('$value'); return parsed != null && parsed.isFinite ? parsed : null; }
// Процент тренда: сервер уже округлил его до целого, но `null` здесь
// значимое значение — «сравнивать не с чем», а не «ноль процентов».
int? _percent(Object? value) { final double? parsed = _double(value); return parsed == null ? null : parsed.round(); }
DateTime? _date(Object? value) => value is String ? DateTime.tryParse(value)?.toLocal() : null;

class ServiceStatus {
  const ServiceStatus({required this.registrationEnabled, required this.maintenance, this.retryAfterSec = 30});
  factory ServiceStatus.fromJson(Map<String, dynamic> json) => ServiceStatus(
    registrationEnabled: json['registrationEnabled'] != false,
    maintenance: json['maintenance'] == true,
    retryAfterSec: (json['retryAfterSec'] == null ? 30 : _int(json['retryAfterSec'])).clamp(1, 3600).toInt(),
  );
  static const available = ServiceStatus(registrationEnabled: true, maintenance: false);
  final bool registrationEnabled;
  final bool maintenance;
  final int retryAfterSec;
}

class GeoPoint {
  const GeoPoint(this.lat, this.lon, {required this.approximate, required this.source, this.country, this.countryCode});
  factory GeoPoint.fromJson(Map<String, dynamic> json) => GeoPoint(
    _double(json['lat']) ?? double.nan, _double(json['lon']) ?? double.nan,
    approximate: json['approximate'] != false,
    source: '${json['source'] ?? ''}', country: json['country'] as String?, countryCode: json['countryCode'] as String?,
  );
  final double lat;
  final double lon;
  final bool approximate;
  final String source;
  final String? country;
  final String? countryCode;
  bool get valid => lat.isFinite && lon.isFinite && lat >= -90 && lat <= 90 && lon >= -180 && lon <= 180;
}

class ActiveTunnelDevice {
  const ActiveTunnelDevice({required this.id, required this.deviceName, required this.platform, required this.lastSeen, required this.isCurrent, required this.connectedAt, required this.durationSec, required this.node, this.origin, this.status = 'ACTIVE', this.sessionId});
  factory ActiveTunnelDevice.fromJson(Map<String, dynamic> json) {
    final originJson = _map(json['origin']);
    return ActiveTunnelDevice(
      status: '${json['status'] ?? 'ACTIVE'}', id: '${json['id'] ?? ''}', deviceName: '${json['deviceName'] ?? 'Device'}', platform: '${json['platform'] ?? ''}',
      // Сессия нужна кнопке «Отключить»: она гасит именно туннель,
      // а не выкидывает устройство из аккаунта.
      sessionId: '${json['sessionId'] ?? ''}'.isEmpty ? null : '${json['sessionId']}',
      lastSeen: _date(json['lastSeen']), isCurrent: json['isCurrent'] == true, connectedAt: _date(json['connectedAt']),
      durationSec: _int(json['durationSec']), node: VpnNodeInfo.fromJson(_map(json['node'])),
      origin: originJson.isEmpty ? null : GeoPoint.fromJson(originJson),
    );
  }
  final String id;
  final String? sessionId;
  final String deviceName;
  final String status;
  final String platform;
  final DateTime? lastSeen;
  final bool isCurrent;
  final DateTime? connectedAt;
  final int durationSec;
  final GeoPoint? origin;
  final VpnNodeInfo node;
}

class ActiveMapSnapshot {
  const ActiveMapSnapshot({required this.serverTime, required this.pollAfterMs, required this.activeTunnels, required this.maxDevices, required this.truncated, required this.service, required this.devices});
  factory ActiveMapSnapshot.fromJson(Map<String, dynamic> json) {
    final raw = json['devices'];
    return ActiveMapSnapshot(
      serverTime: _date(json['serverTime']), pollAfterMs: _int(json['pollAfterMs']).clamp(3000, 60000).toInt(),
      activeTunnels: _int(json['activeTunnels']), maxDevices: _int(json['maxDevices']), truncated: json['truncated'] == true,
      service: ServiceStatus.fromJson(_map(json['service'])),
      devices: raw is List ? raw.map((e) => ActiveTunnelDevice.fromJson(_map(e))).toList(growable: false) : const <ActiveTunnelDevice>[],
    );
  }
  final DateTime? serverTime;
  final int pollAfterMs;
  final int activeTunnels;
  final int maxDevices;
  final bool truncated;
  final ServiceStatus service;
  final List<ActiveTunnelDevice> devices;
}

enum AnalyticsPeriod { day, week, month }
extension AnalyticsPeriodWire on AnalyticsPeriod { String get wire => name; }

class TrafficPoint {
  const TrafficPoint({required this.start, required this.downloadBytes, required this.uploadBytes});
  factory TrafficPoint.fromJson(Map<String, dynamic> json) => TrafficPoint(start: _date(json['start']), downloadBytes: _int(json['downloadBytes']), uploadBytes: _int(json['uploadBytes']));
  final DateTime? start;
  final int downloadBytes;
  final int uploadBytes;
  int get totalBytes => downloadBytes + uploadBytes;
}

class AnalyticsDevice {
  const AnalyticsDevice({required this.deviceId, required this.deviceName, required this.platform, required this.downloadBytes, required this.uploadBytes});
  factory AnalyticsDevice.fromJson(Map<String, dynamic> json) => AnalyticsDevice(deviceId: '${json['deviceId'] ?? ''}', deviceName: '${json['deviceName'] ?? 'Device'}', platform: '${json['platform'] ?? ''}', downloadBytes: _int(json['downloadBytes']), uploadBytes: _int(json['uploadBytes']));
  final String deviceId, deviceName, platform;
  final int downloadBytes, uploadBytes;
  int get totalBytes => downloadBytes + uploadBytes;
}

class DomainUsage {
  const DomainUsage({required this.domain, required this.category, required this.downloadBytes, required this.uploadBytes, required this.connections, this.lastSeenAt, this.faviconUrl});
  factory DomainUsage.fromJson(Map<String, dynamic> json) => DomainUsage(domain: '${json['domain'] ?? ''}', category: '${json['category'] ?? ''}', downloadBytes: _int(json['downloadBytes']), uploadBytes: _int(json['uploadBytes']), connections: _int(json['connections']), lastSeenAt: _date(json['lastSeenAt']), faviconUrl: json['faviconUrl'] as String?);
  final String domain, category;
  final int downloadBytes, uploadBytes, connections;
  final DateTime? lastSeenAt;
  final String? faviconUrl;
  int get totalBytes => downloadBytes + uploadBytes;
}

class CategoryUsage {
  const CategoryUsage({required this.category, required this.downloadBytes, required this.uploadBytes});
  factory CategoryUsage.fromJson(Map<String, dynamic> json) => CategoryUsage(category: '${json['category'] ?? ''}', downloadBytes: _int(json['downloadBytes']), uploadBytes: _int(json['uploadBytes']));
  final String category;
  final int downloadBytes, uploadBytes;
  int get totalBytes => downloadBytes + uploadBytes;
}

class ServiceBudget {
  const ServiceBudget({required this.available, required this.usedBytes, required this.budgetBytes, required this.usedPercent, this.cycleStart, this.cycleEnd, this.lastPolledAt, this.adminOnly = false});
  factory ServiceBudget.fromJson(Map<String, dynamic> json) => ServiceBudget(adminOnly: json['adminOnly'] == true, available: json['available'] == true, usedBytes: _int(json['usedBytes']), budgetBytes: _int(json['budgetBytes']), usedPercent: _double(json['usedPercent']) ?? 0, cycleStart: _date(json['cycleStart']), cycleEnd: _date(json['cycleEnd']), lastPolledAt: _date(json['lastPolledAt']));
  final bool available;
  final bool adminOnly;
  final int usedBytes, budgetBytes;
  final double usedPercent;
  final DateTime? cycleStart, cycleEnd, lastPolledAt;
}

/// Прошлое окно той же длины — основа тренда «+12 %».
///
/// Сервер обрезает его по прошедшей части текущего периода: «сегодня с
/// утра» сравнивается со «вчера до этого же часа». Иначе утро всегда
/// показывало бы падение относительно полных вчерашних суток.
class PreviousWindow {
  const PreviousWindow({required this.start, required this.end, required this.downloadBytes, required this.uploadBytes});
  factory PreviousWindow.fromJson(Map<String, dynamic> json) => PreviousWindow(
    start: _date(json['start']), end: _date(json['end']),
    downloadBytes: _int(json['downloadBytes']), uploadBytes: _int(json['uploadBytes']),
  );
  final DateTime? start, end;
  final int downloadBytes, uploadBytes;
  int get totalBytes => downloadBytes + uploadBytes;
}

/// Тренд период-к-периоду в целых процентах.
///
/// `comparable == false` означает «сравнивать не с чем»: история короче
/// прошлого окна. Рисовать тогда «+100 %» нельзя — это была бы не
/// разница трафика, а дата включения аналитики.
class TrendInfo {
  const TrendInfo({required this.comparable, this.downloadPercent, this.uploadPercent, this.totalPercent});
  factory TrendInfo.fromJson(Map<String, dynamic> json) => TrendInfo(
    comparable: json['comparable'] == true,
    downloadPercent: _percent(json['downloadPercent']),
    uploadPercent: _percent(json['uploadPercent']),
    totalPercent: _percent(json['totalPercent']),
  );

  /// Старый ответ API: полей нет, бейджи не рисуем.
  static const TrendInfo none = TrendInfo(comparable: false);

  final bool comparable;
  final int? downloadPercent, uploadPercent, totalPercent;

  /// Значения для показа: `null` — бейджа нет вовсе.
  int? get download => comparable ? downloadPercent : null;
  int? get upload => comparable ? uploadPercent : null;
  int? get total => comparable ? totalPercent : null;
}

class AnalyticsSnapshot {
  const AnalyticsSnapshot({required this.period, required this.partial, required this.coverageSince, required this.downloadBytes, required this.uploadBytes, required this.series, required this.devices, required this.domainsEnabled, required this.domainWindowDays, required this.domains, required this.categories, required this.budget, this.quota, this.previous, this.trend = TrendInfo.none});
  factory AnalyticsSnapshot.fromJson(Map<String, dynamic> json) {
    final coverage = _map(json['coverage']); final totals = _map(json['totals']); final domains = _map(json['domains']);
    List<T> list<T>(Object? raw, T Function(Map<String,dynamic>) parse) => raw is List ? raw.map((e) => parse(_map(e))).toList(growable:false) : <T>[];
    return AnalyticsSnapshot(
      period: '${json['period'] ?? 'day'}', partial: coverage['partial'] == true, coverageSince: _date(coverage['since']),
      downloadBytes: _int(totals['downloadBytes']), uploadBytes: _int(totals['uploadBytes']),
      series: list(json['series'], TrafficPoint.fromJson), devices: list(json['devices'], AnalyticsDevice.fromJson),
      domainsEnabled: domains['enabled'] == true, domainWindowDays: _int(domains['windowDays']), domains: list(domains['items'], DomainUsage.fromJson),
      categories: list(json['categories'], CategoryUsage.fromJson), budget: ServiceBudget.fromJson(_map(json['budget'])),
      // Личный лимит тарифа — не путать с `budget` выше: тот сервисный
      // и виден только админам.
      quota: json['quota'] == null ? null : QuotaInfo.fromJson(_map(json['quota'])),
      // Старые ответы без previous/trend обязаны разбираться как раньше:
      // клиент обновляется не одновременно с сервером.
      previous: json['previous'] == null ? null : PreviousWindow.fromJson(_map(json['previous'])),
      trend: json['trend'] == null ? TrendInfo.none : TrendInfo.fromJson(_map(json['trend'])),
    );
  }
  final String period;
  final bool partial;
  final DateTime? coverageSince;
  final int downloadBytes, uploadBytes;
  final List<TrafficPoint> series;
  final List<AnalyticsDevice> devices;
  final bool domainsEnabled;
  final int domainWindowDays;
  final List<DomainUsage> domains;
  final List<CategoryUsage> categories;
  final ServiceBudget budget;

  /// Месячный лимит тарифа по данным сервера; `null` — старый ответ API.
  final QuotaInfo? quota;

  /// Прошлое окно той же длины; `null` — старый ответ API.
  final PreviousWindow? previous;

  /// Тренд к прошлому окну; для старых ответов — [TrendInfo.none].
  final TrendInfo trend;
}
